server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -225,6 +227,66 @@ public U next() {
         }
     }
 
+    public static <T, U> Iterator<U> zip(Iterator<? extends T> input, BiFunction<Integer, T, ? extends U> fn) {
```

Comment:
Any particular reason you chose to call this `zip`? I know `zip` as a function that "zips" _two or more_ iterators. This looks more like a `map` iterator to me (but with the index passed). Would it maybe make more sense to rename this to `mapWIthIndex` or something along those lines?

Reply:
These are great points, my memory is flawed. Scala calls this `zipWithIndex`, which is what I was basically wanting.

I went back and forth between `mapWithIndex`, `zipWithIndex` and `enumerate` but I ended up going with `enumerate` in the end since we already have a `map` call, and may at some point want to add a proper `zip` iterator.